# 1 Lists
1. Create a list 'movies' that contains the names of at least six movies you like

In [22]:
movies = ['Spiderman: No Way Home', 'Spririted Away', 'Train to Busan', 'Avenger: Infinity War', 'Avengers: Endgame', 'Your Name']

2. Use slicing, and create a list of three first movies in the list

In [23]:
movies_3 = movies[:3]
movies_3

['Spiderman: No Way Home', 'Spririted Away', 'Train to Busan']

3. Use slicing and list concatenation to create a list of the first two and the last two movies

In [24]:
first_2 = movies[:2]
last_2 = movies[4:]
first_and_last = first_2 + last_2
first_and_last

['Spiderman: No Way Home', 'Spririted Away', 'Avengers: Endgame', 'Your Name']

# 2 Loops
1. Create a list 'numbers' that is the numbers 70 through 79 in the following manner: create an empty list and add numbers to it in a loop (do not use list comprehension or 'list' function here!


In [25]:
numbers = []
for i in range(70, 80):
    numbers.append(i)
numbers

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

2. Use loop to compute sum of all integers from 1 till 100

In [26]:
sum = 0;
for i in range(1, 101):
    sum += i
sum

5050

3. Assign people to seats. Consider names Adam, Ashin, Inukai, Tanaka, and Ikki; and seats 33, 12, 45, 2, and 17. Print seat assignments by assigning each name to the corresponding seat. 

In [27]:
people = ['Adam', 'Ashin', 'Inukai', 'Tanaka', 'Ikki']
seats = [33, 12, 45, 2, 17]
for i in range(len(people)):
    print(people[i],':', seats[i])

Adam : 33
Ashin : 12
Inukai : 45
Tanaka : 2
Ikki : 17


# 3 Comprehensions
1. Use list comprehension to create a list of squares of numbers

In [28]:
[x*x for x in range(1, 11)]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

2. Use dict comprehension to create a dict of numbers and their squares

In [29]:
{x: x*x for x in range(1,11)}

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}

# 4 Functions and data transformations
1. Write a function that takes in time in numeric form of HHMM, and returns it in the numeric form of HH.HH.The fuction should return the result as a number

In [30]:
def convert(x):
    min = x % 100
    min_frac = min / 60
    hours = x // 100
    return hours + min_frac

# 5 Load Data
1. Load the csv data file

In [31]:
import pandas as pd
df = pd.read_csv("../data/gapminder.csv.bz2", sep="\t")

2. Do basic sanity checks:

a. How many variables (columns) is there in the data? Ensure you know the variables in the data. Keep the documentation nearby

In [32]:
df.shape[1]

25

There is a total of 25 columns, including the column that contains all of the indexes.

b. How many rows of data is there?

In [33]:
df.shape[0]

13055

There is a total of 13,055 rows in the dataset.

c. print the first few lines of data. Does it look resonable?

In [34]:
df.head()

iso3   name iso2    region                       sub-region  \
0  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
1  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
2  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
3  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
4  ABW  Aruba   AW  Americas  Latin America and the Caribbean   

  intermediate-region    time  totalPopulation  fertilityRate  lifeExpectancy  \
0           Caribbean  1960.0          54211.0          4.820          65.662   
1           Caribbean  1961.0          55438.0          4.655          66.074   
2           Caribbean  1962.0          56225.0          4.471          66.444   
3           Caribbean  1963.0          56695.0          4.271          66.787   
4           Caribbean  1964.0          57032.0          4.059          67.113   

   ...  accessElectricity  agriculturalLand  agricultureTractors  \
0  ...                NaN               NaN                  NaN   
1  ...                NaN              20.0                  NaN   
2  ...                NaN              20.0                  NaN   
3  ...                NaN              20.0                  NaN   
4  ...                NaN              20.0                  NaN   

   cerealProduction  fertilizerHa        co2  greenhouseGases     co2_PC  \
0               NaN           NaN  11092.675              NaN  204.62037   
1               NaN           NaN  11576.719              NaN  208.82281   
2               NaN           NaN  12713.489              NaN  226.11808   
3               NaN           NaN  12178.107              NaN  214.80037   
4               NaN           NaN  11840.743              NaN  207.61578   

   pm2.5_35  battleDeaths  
0       NaN           NaN  
1       NaN           NaN  
2       NaN           NaN  
3       NaN           NaN  
4       NaN           NaN  

[5 rows x 25 columns]

Yes, it looks reasonable

# 6 Wealth
1. How many different countries are there in the data?

In [35]:
countries = df.name.nunique()
countries

249

There are 249 different countries in the dataset. I used the unique() method to get the number of different values in the name column.

2. What is the earliest and the most recent year in the dataset?

In [36]:
df.time.min()

1960.0

In [37]:
df.time.max()

2019.0

The earliest year is 1960 and the lates year is 2019. I used the min() and max() method on the time column to find the smallest time and the largest time. Smallest would mean earliest and largest would mean latest.

3. For which year do we have the most recent GDP data?
Hint: You can remove all cases where GDP is missing.

In [38]:
ndf = df.dropna(subset=['GDP_PC'])
ndf.time.max()

2019.0

We have the most recent GDP data from the year 2019. I found this by dropping all of the NA values specifically in the GDP_PC column. I did this by specifying the subset to GDP_PC which I learned from the following website: https://towardsdatascience.com/how-to-drop-rows-in-pandas-dataframes-with-nan-values-in-certain-columns-7613ad1a7f25

4. What is the average wealth on this planet as of 2019? Let’s just compute average GDP across
all countries in 2019 and ignore the fact that countries are of different size.

In [39]:
ndf[ndf.time == 2019].GDP_PC.mean()

15055.352217243239

The average wealth on this planet in the year 2019 is 15,055.35. I computed this by filtering the time to 2019 and finding the mean of the GDP_PC column

5. But not all countries may have the same most recent year where GDP data is present. Which 5
countries countries have the largest number of the most recent years missing? (For instance, imagine
there is GDP data for Funan for 2015, 2016, and 2017; but for Khmer Empire only for 2015. Hence
Khmer Empire has two of the most recent years missing.) Till which year do these countries have
data? What do you think, why do these countries have issues with more recent data?

Hint: you may group by country and find max value for the year. In the resulting series, find the
min/max. Check out the nlargest method.

Hint2: two of these countries are Lichtenstein and Faroe Islands.

In [40]:
ndf.groupby("name").time.max().nsmallest()

name
Djibouti                    2010.0
Faroe Islands               2010.0
Liechtenstein               2010.0
Turks and Caicos Islands    2010.0
Eritrea                     2011.0
Name: time, dtype: float64

The 5 countries with the largest number of most recent years missing is Djibouti, Faroe Islands, Liechtenstein, Turks and Caicos Islands, and Eritrea. These countries have GDP data until 2010 and 2011. It is possible that the soure of this dataset was unable to find anything on these 5 countries during the past 10 years. Or maybe, because these are smaller countries, there is some power struggle going on that is preventing them from reporting their GDP_PC.

I grouped by the countries and selected the 5 countries with the oldest, most recent year.

6. Now let’s compare the continents. We’ll make it easy again and just compute the average
wealth (i.e. GDP) for each continent in 2019, and we use region as continent. We disregard the fact
that countries are of different size. Print the continents, and the corresponding GDP in a decreasing
order. Do you think this order is reasonable?
Remember to use only the most recent data!

Hint: check out methods groupby and sort_values.

Hint2: you should see value around 11,800 for Oceania.

In [41]:
ndf[ndf.time == 2019].groupby("region").GDP_PC.mean().sort_values(ascending = False)

region
Europe      33227.983767
Americas    14516.826303
Asia        14116.124163
Oceania     11817.735554
Africa       2765.708956
Name: GDP_PC, dtype: float64

I think this data makes sense. Europe exports a lot of luxury brands (fashion, car, etc.) and doesn't have much underdeveloped areas hwen compared to the rest of the continents. However, I was shocked to see that Europe's average GDP_PC doubles the Americas and Asia.

I did this by filtering by the year 2019 and grouping by continent. Then, I found the mean of the GDP of each continent and sorted it from largest to smallest.

7. But this was just about the average numbers. Now for each continent let’s also find the richest and the poorest country, the corresponding GDP, and population (2019). Note: While this gives a hint about inequality, we still ignore the intra-county inequality. Quite likely the rich in the poor countries earn more than the poor in the rich countries. Unfortunately we cannot tell based on these data.

Hint: if you can find a good solution yourself, go for it! But here is a suggestion how you can do it,
step-by-step.

(a) Check out indexing with .loc and .iloc (McKinney, p 143)

(b) Check out methods .idxmax and .idxmin (McKinney p 158-159)

(c) Find the index of the largest GDP value in data (use .idxmax).

(d) Extract the row of dataframe that corresponds to the index you computed in 7c

(e) Now repeat the two previous steps either with .groupby, or if you cannot figure this out then with a for-loop over all continents.

Alternatively, you can extract the values using construct like data.gdp == data.gdp.min(). You can also just loop over continents, and for each continent find the richest and poorest country as of 2019 (check out methods nlargest and nsmallest).

In [42]:
temp = ndf.rename({'name':'country'}, axis=1)
maxi = temp[temp.time == 2019].groupby('region').GDP_PC.idxmax()

country = []
gdp = []
population = []

for i in range(5):
    country.append(temp.loc[maxi[i]].country)
    gdp.append(temp.loc[maxi[i]].GDP_PC)
    population.append(temp.loc[maxi[i]].totalPopulation)

In [43]:
mini = temp[temp.time == 2019].groupby('region').GDP_PC.idxmin()
for i in range(5):
    country.append(temp.loc[mini[i]].country)
    gdp.append(temp.loc[mini[i]].GDP_PC)
    population.append(temp.loc[mini[i]].totalPopulation)

In [44]:
c = pd.Series(data = country)
g = pd.Series(data = gdp)
p = pd.Series(data = population)

I used idxmax() to find the index of the maximum GDP from each continent. I made 3 lists: country, gdp, and population. I then used a for loop to add the country, gdp, and population into lists by searching for the index in the temo data frame. This is so the index of everything stays the same (ex: country[0] will correspond to gdp[0] and population[0]). I then convert the lists into Series so that I can concat them into dataframes later on. 

8. Finally, let’s put all this into a single data frame. The data frame should contain continent name as index, and for each continent 6 variables: name, GDP and population for the poorest country, and the same for the richest country. Pick suitable variable names!

Hint: consult Python notes 4.6: Combining data into data frames. You can use .set_index to convert a variable into index.

In [45]:
rich_and_poor = pd.concat((g, p), axis=1)
rich_and_poor = rich_and_poor.rename({0:'gdp', 1:'population'}, axis=1)
rich_and_poor = rich_and_poor.set_index(c)

Using Python notes, I used concat to make a dataframe and renamed the columns for easy reading and also set the index to the country names.

9. Comment the list of poorest and richest countries. What do you think about these lists. Did
you know that Bermuda is the richest country in Americas? Do you know why? Why do most of the rich countries have small population?

From richest to poorest: Luxembourg, Bermuda, Qatar, Australia, Seychelles, Ukraine, Solomon Islands, Haiti, Afghanistan, and Burundi. Maybe the reason why the richest countries have small population is because it might be easier to govern and introduce policies. It might also be because the countries still have valuable resources that can be exported.

# 7 Health

1. How many countries do not have LE data for year 1960? How many countries do not have this information for year 2019?

In [47]:
df[df.time == 2019].lifeExpectancy.isnull().sum()
df[df.time == 1960].lifeExpectancy.isnull().sum()

27

27 countries did not have the life expectancy data in the year 1960, and 19 countries don't have the life expectancy data in 2019.

I found this data by filtering the time to 2019, and removing all of rows with nulls in the lifeExpectancy column. And then I used sum() to calculate the number of countries. I used the same process for the year 1960, except I filtered the time to 1960.

2. What is the shortest and longest LE in data? Which years/countries does this correspond to?

Hint: you can use methods nlargest to extract dataframe rows, and idxmax to extract index of
row with the maximum value.

In [48]:
max_le_id = df.lifeExpectancy.idxmax()

In [49]:
min_le_id = df.lifeExpectancy.idxmin()

The longest life expectancy in the data is 85.42 years in the country San Marino during the year 2012. The shortest life expectancy is 18.9 years in the country Cambodia during the year 1977.

I found this by using idxmax() and idxmin(), which gets me the index of the largest and smallest value of a column.

3. If you did this correctly, you notice that the shortest LE is less than 20 years. What historical
events does it correspond to? (You may consult Wikipedia).

This year is in the middle of the Cambodian Genocide.

4. Find the country with longest and shortest LE for each continent.
Hint: you can use the same approach as for GDP by continent.

In [51]:
longest_le = temp.groupby('region').lifeExpectancy.idxmax()
longest = []
for i in range(5):
    longest.append(temp.loc[longest_le[i]].country)
    
l = pd.Series(data = longest)

In [52]:
shortest_le = temp.groupby('region').lifeExpectancy.idxmin()
shortest = []
for i in range(5):
    shortest.append(temp.loc[shortest_le[i]].country)

s = pd.Series(data = shortest)

I used idxmax() and idxmin() to find the index of the largest and smallest values of the lifeExpectancy column. Then, I used a for loop and used the index to find the country and appended the it to a list.

5. Now lets find, for each country, the first valid LE. Find the corresponding year, and LE value.

Hint: you can use the same approach as when finding the largest GDP on each continent: find the index of minimum year of each country, and then extract the rows based on the index.

In [30]:
valid_le = df.dropna(subset=['lifeExpectancy'])
valid_le = valid_le.rename({'name':'country'}, axis = 1)

In [31]:
min_year_index = valid_le.groupby('country').time.idxmin()

min_y = []
countries = []
le = []

for i in range(0, len(min_year_index)):
    min_y.append(valid_le.loc[min_year_index[i]].time)
    countries.append(valid_le.loc[min_year_index[i]].country)
    le.append(valid_le.loc[min_year_index[i]].lifeExpectancy)

min_year = pd.Series(data=min_y)
c = pd.Series(data=countries)
l = pd.Series(data=le)

In [105]:
min_year_le = pd.concat((c, min_year, l), axis = 1)
min_year_le = min_year_le.rename({0:'country', 1:'year', 2:'le'}, axis=1)
min_year_le

I got rid of the rows with nulls in the lifeExpectancy column and renamed the column for better accessability. I used idxmin(), and ran it through a for loop (basically the same thing I did earlier).

6. Now repeat this with the most recent valid LE.

In [32]:
max_year_index = valid_le.groupby('country').time.idxmax()
max_y = []
max_le = []

for i in range(0, len(max_year_index)):
    max_y.append(valid_le.loc[max_year_index[i]].time)
    max_le.append(valid_le.loc[max_year_index[i]].lifeExpectancy)
    
max_year = pd.Series(data=max_y)
max_leS = pd.Series(data=max_le)

In [ ]:
max_year_le = pd.concat((c, max_year, max_leS), axis = 1)
max_year_le = max_year_le.rename({0:'country', 1: 'year', 2:'le'}, axis=1)
max_year_le

I used similar logic as when I computed the maximum GDP per continent

7. Finally, let’s compute the growth rate. Show the countries, growth rates, and the corresponding life expectancies for 10 fastest and 10 slowest growers.

Hint: for each country, compute the first valid year of life expectancy, last valid year of life expectancy, and find their life expectancies for the corresponding years. You can do it in a fairly similar way as in the previous question. Now use this formula for each country. At the end just order the result.

In [100]:
rate = (max_year_le['le'] - min_year_le['le'])/(max_year_le['year'] - min_year_le['year'])
rate = pd.concat((c, rate, max_year, max_leS, min_year, l), axis = 1)
rate = rate.rename({0: 'country', 1: 'growth_rate', 2:'latest_valid_year', 3: 'latest_valid_le', 4:'earliest_valid_year', 5: 'earliest_valid_le'}, axis = 1)

In [102]:
rate.nsmallest(10, 'growth_rate')

country  growth_rate  latest_valid_year  latest_valid_le  \
116  Marshall Islands    -0.530957             2000.0         65.23902   
139             Palau     0.004000             2005.0         69.12927   
191           Ukraine     0.059793             2019.0         71.82732   
101            Latvia     0.092410             2019.0         75.23902   
26           Bulgaria     0.096052             2019.0         74.91463   
107         Lithuania     0.106556             2019.0         76.13415   
103           Lesotho     0.108678             2019.0         54.33100   
15            Belarus     0.110487             2019.0         74.22683   
163        Seychelles     0.111445             2019.0         73.94390   
61               Fiji     0.112424             2019.0         67.44400   

     earliest_valid_year  earliest_valid_le  
116               1987.0           72.14146  
139               1990.0           69.06927  
191               1960.0           68.29954  
101               1960.0           69.78683  
26                1960.0           69.24756  
107               1960.0           69.84732  
103               1960.0           47.91900  
15                1960.0           67.70810  
163               1980.0           69.59756  
61                1960.0           60.81100

In [103]:
rate.nlargest(10, 'growth_rate')

country  growth_rate  latest_valid_year  latest_valid_le  \
113     Maldives     0.704712             2019.0           78.921   
20        Bhutan     0.631373             2019.0           71.777   
201        Yemen     0.613661             2019.0           66.125   
183  Timor-Leste     0.606203             2019.0           69.495   
128        Nepal     0.596525             2019.0           70.778   
137         Oman     0.596424             2019.0           77.861   
187      Tunisia     0.587763             2019.0           76.699   
37         China     0.562492             2019.0           76.912   
0    Afghanistan     0.548932             2019.0           64.833   
188       Turkey     0.547831             2019.0           77.691   

     earliest_valid_year  earliest_valid_le  
113               1960.0             37.343  
20                1960.0             34.526  
201               1960.0             29.919  
183               1960.0             33.729  
128               1960.0             35.583  
137               1960.0             42.672  
187               1960.0             42.021  
37                1960.0             43.725  
0                 1960.0             32.446  
188               1960.0             45.369

Since I knew that I can subtract 2 dataframes of the same length from each other, I decided to specify the columns that I wanted to subtract and computed the growth rate by using the following formula: (y2-y1)/(x2-x1). I then used concat() to combine the country, growth rate, years, and life expectancies into 1 dataframe and sorted it by growth rate.
8. Do you see a pattern (or multiple patterns) here? Remember: you are looking at growth of
life expectancy over an extended period.

Generally, countries are almost showing a growth in life expectancy. The only exceptiona are Marshall Islands, Cayman Islands, and San Marino. This is because there is only one year with data about life expectancy for Cayman Islands and San Marino. As for Marshall Islands, there is only 3 years that displays data about life expectancy, in other words, the data is inconsistent. It seems like the countries with the highest growth rates tend to be in Asia and the countries with the lowest growth rates ten to be in Europe, Africa, and Oceania.

# How much time did you spend?
And finally-finally, tell us how much time (how many hours) did you spend on this PS!

I think I spent at least 24 hours of coding on this project. I'm not entirely sure of the exact number of hours, but I know I spent at least 24 hours coding this assignment.

# References
Lubanovic, B. (2014) Introducing Python: Modern Computing in Simple Packages, O’Reilly Media.